### PR TITLE
Add support for RDTSC and RDTSCP

### LINKED ISSUE #10: Missing support for RDTSC / RDTSCP
While these are categorized as "system" instructions, they are also popular within userspace applications. While it is unlikely that microx can provide a "correct" implementation, it can at least provide a way for the user to provide an implementation.

Something like:
```
@@ -1347,6 +1421,18 @@ static bool Emulate(const Executor *executor, uintptr_t &next_pc,
       WriteGPR(reg0, mem0);
       return true;

+    case XED_IFORM_RDTSC:
+    case XED_IFORM_RDTSCP: {
+        Data tsc;
+        if (!executor->ReadTsc(tsc)) {
+          status = ExecutorStatus::kErrorReadTsc;
+          return false;
+        }
+        WriteGPR(reg0, *reinterpret_cast<uint32_t *>(&tsc.bytes[0]));
+        WriteGPR(reg1, *reinterpret_cast<uint32_t *>(&tsc.bytes[4]));
+      }
+      return true;
+
     default:
       return false;
   }
@@ -1382,6 +1468,14 @@ static bool UsesUnsupportedAttributes(void) {

 static bool UsesUnsupportedFeatures(const Executor *executor) {
   switch (xed_decoded_inst_get_category(gXedd)) {
+    case XED_CATEGORY_SYSTEM:
+      switch (xed_decoded_inst_get_iform_enum(gXedd)) {
+        case XED_IFORM_RDTSC:
+        case XED_IFORM_RDTSCP:
+          return false;
+        default:
+          return true;
+      }
     case XED_CATEGORY_3DNOW:
     case XED_CATEGORY_MPX:
     case XED_CATEGORY_AES:
```

### PR DESCRIPTION
**Pull request by awesie**
Fixes #10.

Instead of adding another method for reading the TSC, just use ReadReg. XED already treats TSC and TSCAUX as registers and these will get correctly read by ReadRegisters, provided that WidestRegister is fixed so that it does not turn TSC into INVALID.

This also adds a simple implementation of TSC by incrementing it by max_num_instructions after execute. A user can hook read_register / write_register if they want to do something different with TSC.

### CODE REVIEW

File: microx/Executor.cpp [hunk 1]
```diff
@@ -1347,6 +1352,15 @@ static bool Emulate(const Executor *executor, uintptr_t &next_pc,
       WriteGPR(reg0, mem0);
       return true;
 
+    case XED_IFORM_RDTSCP:
+        WriteGPR(reg2, ReadValue<uint32_t>(XED_REG_TSCAUX));
+    case XED_IFORM_RDTSC: {
```
Comment (pgoodman) [inline microx/Executor.cpp:1358]: Add explicit fall-through comment, or `[[clang::fallthrough]];`

File: microx/Executor.cpp [hunk 2]
```diff
@@ -907,6 +911,7 @@ static bool Emulate(const Executor *executor, uintptr_t &next_pc,
   const auto stack_reg = WidestRegister(executor, XED_REG_ESP);
   const auto reg0 = xed_decoded_inst_get_reg(gXedd, XED_OPERAND_REG0);
   const auto reg1 = xed_decoded_inst_get_reg(gXedd, XED_OPERAND_REG1);
+  const auto reg2 = xed_decoded_inst_get_reg(gXedd, XED_OPERAND_REG2);
```
Comment (pgoodman) [inline microx/Executor.cpp:914]: Good catch.

File: microx/__init__.py [hunk 3]
```diff
@@ -495,6 +495,10 @@ def execute(self, thread, max_num_instructions=1):
         self._thread = thread
         try:
             super(Process, self).execute(max_num_instructions)
+
+            # Approximate TSC as 1 cycle / instruction.
+            tsc = self.read_register("TSC", thread.REG_HINT_NONE)
```
Comment (pgoodman) [inline microx/__init__.py:500]: You could feasibly make a new reg hint type, but this is fine as well.

**Review by pgoodman — CHANGES_REQUESTED** [review]

**Comment by awesie** [comment]
I added an explicit comment for the fall-through. I couldn't decide how compatible the various options for doing the fall-through attribute were.

**Review by pgoodman — APPROVED** [review]

**Pull request merged**